packages/core-data/src/queried-data/get-query-parts.js

```diff
@@ -0,0 +1,64 @@
+/**
+ * Internal dependencies
+ */
+import { withWeakMapCache } from '../utils';
+
+/**
+ * An object of properties describing a specific query.
+ *
+ * @typedef {WPQueriedDataQueryParts}
+ *
+ * @property {number} page      The query page (1-based index, default 1).
+ * @property {number} perPage   Items per page for query (default 10).
+ * @property {string} stableKey An encoded stable string of all non-pagination
+ *                              query parameters.
+ */
+
+/**
+ * Given a query object, returns an object of parts, including pagination
+ * details (`page` and `perPage`, or default values). All other properties are
+ * encoded into a stable (idempotent) `stableKey` value.
+ *
+ * @param {Object} query Optional query object.
+ *
+ * @return {WPQueriedDataQueryParts} Query parts.
+ */
+export function getQueryParts( query ) {
+	/**
+	 * @type {WPQueriedDataQueryParts}
+	 */
+	const parts = {
+		stableKey: '',
+		page: 1,
+		perPage: 10,
+	};
+
+	// Ensure stable key by sorting keys. Also more efficient for iterating.
+	const keys = Object.keys( query ).sort();
+
+	for ( let i = 0; i < keys.length; i++ ) {
+		const key = keys[ i ];
+		const value = query[ key ];
+
+		switch ( key ) {
+			case 'page':
+			case 'perPage':
+				parts[ key ] = value;
+				break;
+
+			default:
+				// While it's not required to be one, for simplicity's sake
+				// mimic querystring encoding for stable key.
+				parts.stableKey += (
+					( parts.stableKey ? '&' : '' ) +
+					encodeURIComponent( key ) +
+					'=' +
+					encodeURIComponent( value )
```

Comment:
Does it assume that `value` can by only represented by primitive types?

Otherwise, we should add deep sorting for arrays or objects and serialization mechanism.

Reply:
Example test case:
```js
	it( 'encodes stable string key when using arrays', () => {
		const first = getQueryParts( { '?': '&', b: [ 3, 2 ] } );
		const second = getQueryParts( { b: [ 2, 3 ], '?': '&' } );

		expect( first ).toEqual( second );
		expect( first ).toEqual( {
			page: 1,
			perPage: 10,
			stableKey: '%3F=%26&b=2',
		} );
	} );
```
Not sure how stable key should look like in such case.

Reply:
Yeah, currently I think this would only support primitive values. It's not quite so straight-forward, but I could imagine recursing into `getQueryParts` for non-primitive types.

Reply:
>Does it assume that `value` can by only represented by primitive types?

Added deep value support (via `addQueryArgs`) in 0368719.